doc/sphinx/language/core/records.rst

```diff
@@ -15,188 +15,226 @@ synonym :cmd:`Structure`.
+      recursive. To define recursive records, use the
+      :cmd:`Inductive` and :cmd:`CoInductive` commands, resulting in an inductive or coinductive record.
+      Definition of mutually inductive or coinductive records are also allowed, as long
+      as all of the types in the block are records.
+
+   .. note:: Induction schemes are automatically generated for inductive records.
+      Automatic generation of elimination schemes for non-recursive records
+      defined with the :cmd:`Record` command can be activated with the
+      :flag:`Nonrecursive Elimination Schemes` flag (see :ref:`proofschemes-induction-principles`).
+
+   .. todo PR: below: huh?
+
+   .. note:: Records exist in two flavors. In the first,
+      a record :n:`@ident` with parameters :n:`{* @binder }`,
+      constructor :n:`@ident__0`, and fields :n:`{* @name @field_type }`
+      is represented as a variant type with a single
+      constructor: :n:`Variant @ident {* @binder } : @sort := @ident__0
+      {* ( @name @field_type ) }` and projections are defined by case analysis.
+      In the second implementation, records have
+      primitive projections: see :ref:`primitive_projections`.
+
 Constructing records
-~~~~~~~~~~~~~~~~~~~~~~
+~~~~~~~~~~~~~~~~~~~~
+
+   .. insertprodn term_record field_def
 
    .. prodn::
       term_record ::= %{%| {*; @field_def } {? ; } %|%}
       field_def ::= @qualid {* @binder } := @term
 
-   Objects in record types can be written using either an *applicative
-   syntax* of the form :n:`@term_application` or a *record syntax* of
-   the form :n:`@term_record`.
+   Instances of record types can be constructed using either *record syntax*
+   (:n:`@term_record`, shown here) or with *applicative syntax* (see :n:`@term_application`)
+   using the constructor.
 
-   In the applicative form, it takes the
-   form :n:`@qualid {? @univ_annot } {+ @arg }` or
-   :n:`@@qualid {?  @univ_annot } {+ @term1 }` where :n:`@qualid`
-   refers to the constructor. That is, the constructor is seen as a
-   function and the instances for the abstract fields are given in
-   order as arguments of the constructor.
+   .. todo PR: what does it mean if you provide a qualid in field_def that's not just an ident?
+      Is that just to pick out a field defined in another section?
 
-   In the record form, the
-   instance of an abstract field is given as body of the
-   corresponding field name and the instances can be given in any
-   order. If a field is omitted, its instance is implicitly considered
-   to be the :n:`_` placeholder. It can then be filled automatically
-   by unification or by using obligations (see :ref:`programs`).
+   In the record form, the fields can be given in any order.  Fields that can be
+   inferred by unification or by using obligations (see :ref:`programs`) may be omitted.
 
-.. FIXME: move this to the main grammar in the spec chapter
+   .. todo PR: what is the difference between "mkRat ..." and "@mkRat ..."?
```

Comment:
There was no description of `@`, so I added one.  But why isn't the second line in the grammar below `@ term1 {+ term1 }`?  Is that a bug or a feature?

![image](https://user-images.githubusercontent.com/1253341/153126449-9f11d4d4-7720-4f60-a3b0-7261496a5393.png)

Reply:
>  Is that a bug or a feature?

Implicit arguments are attached to global names, so the prefix `@` applies only to global names.

This could somehow be generalized but that's another story... Currently, it applies only to global names (i.e. `@qualid`).

Reply:
The grammar here seems more complex than is necessary.  It would be easier to understand if we didn't have both `{+ arg }` and `{+ term1 }` in parallel productions.  I'd suggest simplifying `term_application` to just `{? @ } @term1 {+ @arg }`.

Given that `@` is a no-op for functions that don't have implicits, it shouldn't be a problem to replace `qualid_annotated` with `term1`.  For the first two lines of `arg`, Coq could give a runtime error pretty simply if the `term1` is not a `qualid_annotated`.

WDYT?

Reply:
I think something like that was tried and it conflicted with uses of `@` as an infix symbol. The conclusion was something like that it would work by adopting a convention that for `@` to be used in the middle of a notation, it has either to be surrounded by two spaces or by no spaces at all, and when used to deactivate implicit arguments to be used with no space.